### DISCUSSION TITLE
Async support?

### DISCUSSION THREAD

**Discussion by rushkii** [python-pillow/Pillow#5831] [Q&A] [answered]
Do you have any plan to make another Pillow with async support?

**Comment by radarhere** [comment, marked as answer]
Hi. There are no current plans.

I would have thought that the end user could just wrap calls to Pillow within their own async code.

Can you give an example of something that is better implemented within Pillow, rather than the end user's code?

**Reply by rushkii** [reply]
Maybe you can see this https://github.com/pyrogram/pyrogram/blob/master/pyrogram/sync.py, it is wrapped all functions and can use both async and sync when calling a function like this:
```python
def just_test():
    call_from_wrapped()

async def just_test():
    await call_from_wrapped()
```
When calling the function in an async function it uses await, and the opposite.

**Reply by radarhere** [reply]
Sure, we could go through and do that, but it would be a large change. Could you give a specific example to demonstrate why this would be helpful?

**Reply by cdgriffith** [reply]
A pretty straightforward example would be use in very `asyncio` heavy libraries like FastAPI.  When using `pillow` to create thumbnails, would be great to simply be able to `await` the `open` / `save` tasks. Currently they have to be put into a background thread so they don't block the webserver.

**Reply by radarhere** [reply]
I've not experimented with asyncio myself, so bear with me - but if you want to `await` an `open()`, is it not as simple as this?
```python
from PIL import Image
import asyncio

async def image_open(path):
    return Image.open(path)

async def main():
    im = await image_open("hopper.png")
    print(im)

asyncio.run(main())
```

**Reply by wiredfool** [reply]
Yes, but it would suck. Especially when the deferred load gets called. 

My understanding of async is that you might yield control to alternate "threads" anywhere there is an await, and that what runs within that async will be scheduled "sometime" when something else potentially yields. 

To make this work, you basically need async all the way down, otherwise you're going to wind up blocking everyone else when you do something that takes a significant amount of time, like waiting for disk IO in Image.open, or pixel unpacking in load, or encoding in save. Async IO all the way down is going to be difficult, as we don't always control the IO primitives.  If your image op takes 100ms and you've done this in an async context without some form of worker thread, then you're going to have a 100ms latency pause in all of the other concurrent operations. 

So essentially, we'd have to setup a worker thread pool for everything that will potentially take time and ensure that it's used for those operations.

**Reply by PogoDigitalism** [reply]
A bit late but, I have an open repository called [PillowInAsync ](https://github.com/PogoDigitalism/PillowInAsync). 

(check the examples folder for diffferent applications of the technique)

I use a thread pool executor with Async IO tasks to make my Pillow image generation work with Async projects such as discord.py. 👍 

I have now added a user-friendly copy-paste class to [SyncInAsync](https://github.com/PogoDigitalism/SyncInAsync) to run synchronous functions in async environments!

**Reply by aclark4life** [reply]
Nice!
